hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/functional/TestHoodieBackedMetadata.java

```diff
@@ -1108,6 +1119,259 @@ public void testMetdataTableCommitFailure() throws Exception {
+    // bootstrap w/ few commits, but rollback one of the commit before bootstrapping.
+    testBootstrap(testTable,true);
+  }
+
+  private void testBootstrap(HoodieTestTable testTable, boolean addRollback) throws Exception {
+
+    // bootstrap w/ 3 or 5 commits
+    testTable.doWriteOperation(testTable, "001", WriteOperationType.INSERT, Arrays.asList("p1", "p2"), Arrays.asList("p1", "p2"),
+        2, true);
+    testTable.doWriteOperation(testTable, "002", WriteOperationType.INSERT, Collections.emptyList(), Arrays.asList("p1", "p2"),
+        2, true);
+    syncAndValidate(testTable);
+
+    if (addRollback) {
+      doRollback(testTable, "003", "004", Collections.singletonList("p3"), Arrays.asList("p1","p2", "p3"), 2);
+    }
+    testTable.doWriteOperation(testTable, "005", WriteOperationType.INSERT, Collections.emptyList(), Arrays.asList("p1", "p2"),
+        4);
+    syncAndValidate(testTable);
+
+    // trigger an upsert and validate
+    testTable.doWriteOperation(testTable, "006", WriteOperationType.UPSERT, Collections.singletonList("p3"),
+        Arrays.asList("p1", "p2", "p3"), 4, false);
+    syncAndValidate(testTable);
+  }
+
+  private void doRollback(HoodieTestTable testTable, String commitTimeToRollback, String commitTime,
+                          List<String> newPartitionsToAdd, List<String> partitionsToAddFiles, int numFilesPerPartition) throws Exception {
+    // trigger an UPSERT that will be rolled back
+    Pair<HoodieCommitMetadata, PartitionFileInfoMap> commitMeta = testTable.doWriteOperation(testTable, commitTimeToRollback, WriteOperationType.UPSERT,
+        newPartitionsToAdd,
+        partitionsToAddFiles, numFilesPerPartition, false);
+    syncTableMetadata();
+
+    // rollback last commit
+    Map<String, List<String>> partitionFilesToDelete = getPartitionFilesToDelete(commitMeta.getKey());
+    HoodieRollbackMetadata rollbackMetadata = testTable.getRollbackMetadata(commitTimeToRollback, commitTime, partitionFilesToDelete);
+    testTable.addRollback(commitTime, rollbackMetadata);
+
+    // delete the resp files from test table before validation
+    for (Map.Entry<String, List<String>> entry : partitionFilesToDelete.entrySet()) {
+      testTable.deleteFilesInPartition(entry.getKey(), entry.getValue());
+    }
+    syncAndValidate(testTable);
+  }
+
+  /**
+   * Test few table operations like insert, upsert, compaction, clean.
+   * @param tableType
+   * @throws Exception
+   */
+  @ParameterizedTest
+  @EnumSource(HoodieTableType.class)
+  public void testTableOperationsWithTestTable(HoodieTableType tableType) throws Exception {
+    init(tableType);
+    HoodieTestTable testTable = HoodieTestTable.of(metaClient);
+    testTableOperations(testTable,false);
+  }
+
+  /**
+   * 1. Enable metadata to sync and validate.
+   * 2. Disable metadata and add few writes to table.
+   * 3. Enable back again to sync and validate.
+   * @throws Exception
+   */
```

Comment:
`@throws Exception` looks redundant here. most of the time we just let exception throw and investigate the failure.

Reply:
if test logic is encapsulate in well-design util APIs, we may not need extra javadoc to explain the flow. Some inline comments might still be helpful but ideally code itself should be able to explain it pretty well